ocaml/fstar-lib/FStar_Parser_LexFStar.ml

```diff
@@ -686,6 +696,22 @@ match %sedlex lexbuf with
    uninterpreted_blob snap name pos buffer lexbuf
  | _ -> assert false
 
+and use_lang_blob snap name pos buffer lexbuf =
+match %sedlex lexbuf with
+ | "#end-lang" ->
```

Comment:
What do you think about ending an extension block with `#lang-fstar` (or `#lang-pulse` for that matter)?

I wonder if it is feasible for the extension parser/lexer to call back into the main parser.  That is, if we could just add a `#lang-` rule to the pulse grammer?

One disadvantage of detecting the end of the extension block here is that it is unaware of the lexical syntax of the extension language (namely string literals and comments).

```fstar
#lang-pulse

(*
We can exit a pulse block like this:
#end-lang
*)

fn foo () requires emp returns _: string ensures emp {
  "This is how you end a pulse extension block:
#end-lang
Everything after that is F* again."
}
```

Reply:
Or an optional #end-lang-X to match a #lang-X?

The extension parser for Pulse does call back into the F* parser and lexer. So, it can, in principle, handle a nested ```#lang-x ... #end-lang-x``` block

The disadvantage you mentioned remains, but I guess it's not so bad ... you should be aware of the placement of the end-lang block.

That said, for Pulse itself, I don't see any reason to actually end the #lang-pulse block. Do you?

Reply:
What I'm suggesting is that `#lang-foo` means "parse the *rest of the file* using Foo."  And then the parser for Foo could have the same command, calling back to the main parser.

```fstar
#lang-pulse

fn foo () requires emp ensures emp { () }

// not really necessary, as you point out
#lang-fstar

let bar = 42
```

Reply:
Right, this works. You can even do

```
#lang-pulse
fn stuff ...
#lang-pulse
fn more stuff
```

And it will recursively call back into the Pulse parser for the second block

Reply:
I also registered an #lang-fstar parser and removed #end-lang